jwst/skymatch/skymatch_step.py

```diff
@@ -206,14 +206,18 @@ def _imodel2skyim(self, image_model):
         return sky_im
 
     def _set_sky_background(self, image, sky, step_status):
+
         if self._is_asn:
-            image = datamodel_open(image)
+            dm = datamodel_open(image)
+        else:
+            dm = image
 
         if step_status == "COMPLETE":
-            image.meta.background.method = str(self.skymethod)
-            image.meta.background.level = sky
-            image.meta.background.subtracted = self.subtract
-        image.meta.cal_step.skymatch = step_status
+            dm.meta.background.method = str(self.skymethod)
+            dm.meta.background.level = sky
+            dm.meta.background.subtracted = self.subtract
+        dm.meta.cal_step.skymatch = step_status
 
         if self._is_asn:
-            image.close()
+            dm.save(image)
```

Comment:
Is this actually updating the input cal files on disk? If so, that means folks will need to be careful to keep original versions of their cal files somewhere, for fear of getting overwritten. Can't the results just be passed in memory to the next pipeline steps?

Reply:
If we can keep all the files in memory, then what was the point of trying to reduce the memory usage in https://github.com/spacetelescope/jwst/pull/6874.

We have discussed previously (offline) that https://github.com/spacetelescope/jwst/pull/6874 makes sense only when inputs are lists of file names (i.e., ASN). When all input files are already loaded in memory, then the optimization from https://github.com/spacetelescope/jwst/pull/6874 and changes to the `ModelContainer` not to open input files are almost pointless (this step does not use too much extra memory compared to just input files, maybe extra 12% for boolean masks). What I want to say is that let's say we want to match sky in 1000 input images provided as file names and the images are so large that we cannot load them all at once. Then, the optimization from https://github.com/spacetelescope/jwst/pull/6874 would open/load only two files at a given time.

If input images are file names, would it be a problem to update them? Or do you want to save them under different file name?

Reply:
Ah, and if `skysub` is true, then not only `meta` is modified but data arrays as well.

Reply:
Just trying to understand new flow of things based on #6874 and this PR. In the case of "normal" pipeline processing, where the entire calwebb_image3 pipeline is run start to finish, using an ASN table as the initial input, I'm assuming that once the first step in the pipeline executes (tweakreg in this case) all data are then passed on as an already open `ModelContainer` (right?). Hence in that scenario, the input to `skymatch` is a `ModelContainer` and if so, then the changes implemented in #6874 have no effect and all images are still held in memory during `skymatch` execution? And any updates (e.g. keywords giving the computed "sky" values) are present in the data models that are then passed on to the next step (`outlier_detection` in this case) and no intermediate products are saved to disk that contain those `skymatch` updates. Is this correct so far?

Reply:
Yes. That is correct. For the whole memory saving thing to make sense, input files should not have been loaded into memory (at once). If they are (there is enough memory to do this), it makes no sense to write these files to disk and load them one at a time, (just for fun).

As we have discussed over slack, this memory saving stuff will need some changes to how pipeline works (i.e., passing unopened files between steps). My work in #6874 is in preparation of those changes. The is, I updated code to support limited memory usage, but it makes sense only under certain circumstances (ie., ability to load data on demand and not getting a list of models already loaded).

Reply:
OK, thanks for the explanation. I think I'm all caught up now.

So with regards to this PR, when the modified dm's get saved, they should get saved to files with a different product suffix, so that the original input "cal" files don't get overwritten. Let's use the same "bsub" suffix that's used for other types of background-subtracted products (I know this isn't truly a background subtraction, but it's very similar, and I don't want to throw a new product type into all the DMS subsystems). So instead of just doing a simple `dm.save(image)` it would require use of `.save_model` with appropriate file name suffix specified, such as the way it's done here in the level-2b background step:
https://github.com/spacetelescope/jwst/blob/master/jwst/background/background_step.py#L97
Just use the suffix "bsub".